# Correlation test

Pearson's correlation assumes both variables are continuous and normally distributed!
In summary, while Pearson's correlation can be used for encoded categorical columns under certain conditions (specifically, ordinal data with an encoding that reflects its natural order). we can make a plot and only look at the ordinal columns.

Chi-Square Test of Independence: This is a common method to test the relationship between two categorical variables. It checks whether the distribution of sample categorical data matches an expected distribution.


Spearman's correlation is ideal for ordinal data!

# Chi Squared Test
Does household income affect travel frequency? Does household income have correlation with Education?/with Age?

In [46]:
df=pd.read_csv('/Users/xinqiaoyang/Desktop/coded.csv',sep=',') # Here we use the coded version of the orginal data with missing values since imputed data could be not precise.

In [47]:
df.shape

(1040, 27)

In [48]:
import pandas as pd
from scipy.stats import chi2_contingency

# Create a contingency table
contingency_table = pd.crosstab(df['Household Income'], df['How often do you travel by plane?'])

#print(contingency_table)
# Perform the chi-square test
chi2, p, dof, expected = chi2_contingency(contingency_table)

# Output the results
print(f"Chi-Square Statistic: {chi2}")
print(f"P-value: {p}")
print("Degrees of freedom:", dof)
print("Expected frequencies:\n", expected)
# Interpretation of the results
if p_value < 0.05:  # Common threshold for statistical significance
    print("There is a significant association between Household Income and flying frequency.")
else:
    print("There is no significant association between Household Income and flying frequency.")


Chi-Square Statistic: 94.1393193866575
P-value: 4.2933924080441603e-13
Degrees of freedom: 16
Expected frequencies:
 [[ 15.68    61.985   19.845    0.245    0.245 ]
 [ 25.28    99.935   31.995    0.395    0.395 ]
 [ 46.4    183.425   58.725    0.725    0.725 ]
 [ 24.16    95.5075  30.5775   0.3775   0.3775]
 [ 16.48    65.1475  20.8575   0.2575   0.2575]]
There is a significant association between Household Income and flying frequency.


In [49]:
contingency_table = pd.crosstab(df['Household Income'], df['Education'])

# Now perform the Chi-squared test
chi2, p_value, dof, expected = stats.chi2_contingency(contingency_table)

# Print the results
print("Chi-squared statistic:", chi2)
print("P-value:", p_value)
print("Degrees of freedom:", dof)
print("Expected frequencies:\n", expected)

# Interpretation of the results
if p_value < 0.05:  # Common threshold for statistical significance
    print("There is a significant association between Household Income and Education.")
else:
    print("There is no significant association between Household Income and Education.")


Chi-squared statistic: 111.26742617415329
P-value: 2.574509546577182e-16
Degrees of freedom: 16
Expected frequencies:
 [[ 0.48058252  9.37135922 28.59466019 33.76092233 26.79247573]
 [ 0.77184466 15.05097087 45.92475728 54.22208738 43.03033981]
 [ 1.41747573 27.6407767  84.33980583 99.5776699  79.02427184]
 [ 0.77184466 15.05097087 45.92475728 54.22208738 43.03033981]
 [ 0.55825243 10.88592233 33.21601942 39.21723301 31.12257282]]
There is a significant association between Household Income and Education.


In [50]:
contingency_table = pd.crosstab(df['Household Income'], df['Age'])

# Now perform the Chi-squared test
chi2, p_value, dof, expected = stats.chi2_contingency(contingency_table)

# Print the results
print("Chi-squared statistic:", chi2)
print("P-value:", p_value)
print("Degrees of freedom:", dof)
print("Expected frequencies:\n", expected)

# Interpretation of the results
if p_value < 0.05:  # Common threshold for statistical significance
    print("There is a significant association between Household Income and Age.")
else:
    print("There is no significant association between Household Income and Age.")


Chi-squared statistic: 74.38373026601481
P-value: 4.80377609002487e-11
Degrees of freedom: 12
Expected frequencies:
 [[21.21428571 25.16949153 28.76513317 23.85108959]
 [34.07142857 40.42372881 46.19854722 38.3062954 ]
 [63.         74.74576271 85.42372881 70.83050847]
 [34.07142857 40.42372881 46.19854722 38.3062954 ]
 [24.64285714 29.23728814 33.41404358 27.70581114]]
There is a significant association between Household Income and Age.


Problems:
- Missing data
- specious Causation 
- No info about quality of fit

# Spearman Correlation

Unlike Pearson's correlation, does not measure linear correlation directly. Instead, it assesses how well the relationship between two variables can be described using a monotonic function. In simpler terms, it measures the strength and direction of a monotonic relationship between two variables.



In [51]:
import pandas as pd
from scipy.stats import spearmanr

df = df.dropna()#Notice spearman correlation can't deal with nas, we have to drop it.
spearman_corr, p_value = spearmanr(df['Household Income'], df['How often do you travel by plane?'])


print(f"Spearman Correlation Coefficient: {spearman_corr}")
print(f"P-value: {p_value}")
# Interpret the results
if p_value < 0.05:  # Common threshold for statistical significance
    print("There is a significant association between Household Income and Education.")
else:
    print("There is no significant association between Household Income and Education.")



Spearman Correlation Coefficient: 0.2197029101707585
P-value: 7.177685919792066e-08
There is a significant association between Household Income and Education.


In [52]:
# Perform Spearman's rank correlation test
df = df.dropna()
spearman_correlation, p_value = stats.spearmanr(df['Household Income'], df['Education'])

# Print the results
print("Spearman's rank correlation coefficient:", spearman_correlation)
print("P-value:", p_value)

# Interpret the results
if p_value < 0.05:  # Common threshold for statistical significance
    print("There is a significant association between Household Income and Education.")
else:
    print("There is no significant association between Household Income and Education.")


Spearman's rank correlation coefficient: 0.2887796820227929
P-value: 8.907859661113013e-13
There is a significant association between Household Income and Education.


In [53]:
# Perform Spearman's rank correlation test
spearman_correlation, p_value = stats.spearmanr(df['Household Income'], df['Age'])

# Print the results
print("Spearman's rank correlation coefficient:", spearman_correlation)
print("P-value:", p_value)

# Interpret the results
if p_value < 0.05:  # Common threshold for statistical significance
    print("There is a significant association between Household Income and Age.")
else:
    print("There is no significant association between Household Income and Age.")

Spearman's rank correlation coefficient: 0.23369361003509936
P-value: 9.512887429126508e-09
There is a significant association between Household Income and Age.


Problem:

sample size largely dropped down after dropping all the rows which at least contains 1 missing value.

# Pearson Correlation plot

It measures the linear correlation between pairs of attributes in the dataframe.
The Pearson correlation coefficient ranges from -1 to 1, where:

1 implies a perfect positive linear relationship,
-1 implies a perfect negative linear relationship, and
0 implies no linear correlation.



In [55]:
df=pd.read_csv('/Users/xinqiaoyang/Desktop/coded.csv',sep=',') #read the file again because earlier we dropped the nas.
corr = df.corr(numeric_only=True)
corr.style.background_gradient(cmap='coolwarm') #we can make a plot and only look at the ordinal columns!

With the samll p-values but also small correlation coefficients could be the reason that two variables does not have a big linear correlation but non-linear one.

In [58]:
# For Spearman's rank correlation we also visualize the plots, but also take care that it only applies to "ordinal data" too!
spearman_corr = df.corr(method='spearman')
spearman_corr.style.background_gradient(cmap='coolwarm')

In [60]:
#We can creat a difference correlation plot too to check nonlinear relationship and a look at the very different sections.
dis=corr-spearman_corr 
dis.style.background_gradient(cmap='coolwarm')

"Have you ever smoked a cigarette in an airplane bathroom when it was against the rules? " with "How often do you travel by plane?", this correlation term changed around 15%, but people barely would be honest to this question.

"Is it trude to ask someone to switch seats with you in order to be closer to family?" with "How often do you travel by plane?" , this correlation term changed around 6%.